File: apps/web/test/world-coords.test.ts
/**
 * ⭐⭐ «Kendi diyarıma dön» — 2026-08-16 canlı hatasının hesabı.
 *
 * Kullanıcı bildirdi: *"Dünya sayfasını bir casusluk raporu üzerinden açarsak bu butona
 * tıklayınca aktif şehrimizin diyarına değil dünya sayfasının açıldığı ilk diyara gidiyor."*
 *
 * Sebep: düğme yalnız seçimi boşaltıyordu (`setSel(null)`) ama öncelik zinciri
 * `sel → adres → ev` olduğu için sıra eve değil ADRESE düşüyordu.
 */
import { describe, expect, it } from 'vitest';
import { homeAction, visibleCoords, type Coords } from '../src/lib/world-coords.ts';

const EV: Coords = { k: 3, d: 12 };
const RAPOR: Coords = { k: 1, d: 5 };

describe('koordinat önceliği', () => {
  it('elle seçim her şeyi ezer', () => {
    expect(visibleCoords({ k: 7, d: 99 }, RAPOR, EV)).toEqual({ k: 7, d: 99 });
  });

  it('seçim yoksa ADRES gelir — derin bağlantı hedefi göstermeli', () => {
    expect(visibleCoords(null, RAPOR, EV)).toEqual(RAPOR);
  });

  it('adres de yoksa aktif şehrin diyarı', () => {
    expect(visibleCoords(null, null, EV)).toEqual(EV);
  });

  it('hiçbiri yoksa 1:1', () => {
    expect(visibleCoords(null, null, null)).toEqual({ k: 1, d: 1 });
  });
});

describe('⭐⭐ «Kendi diyarıma dön» düğmesi', () => {
  /**
   * ⚠️⚠️ HATANIN KENDİSİ. Eski davranış `sel = null` ile yetiniyordu; aşağıdaki ilk `expect`
   * o hâlde `RAPOR` döndürüyordu — yani oyuncu kendi diyarına gidemiyordu.
   */
  it('casusluk raporundan açılmış sayfada EVE götürür, adresteki diyara değil', () => {
    const eylem = homeAction(RAPOR);
    expect(eylem.clearUrl).toBe(true);

    // Adres temizlendiği için zincirde `fromUrl` artık yok.
    const sonra = visibleCoords(eylem.sel, null, EV);
    expect(sonra).toEqual(EV);
    expect(sonra).not.toEqual(RAPOR);
  });

  it('⚠️ adres temizlenmezse ESKİ HATA geri gelir — regresyonun kanıtı', () => {
    const eylem = homeAction(RAPOR);
    // `clearUrl` yok sayılsaydı zincir yine adrese düşerdi:
    expect(visibleCoords(eylem.sel, RAPOR, EV)).toEqual(RAPOR);
  });

  it('adres zaten yokken gereksiz yönlendirme YAPILMAZ', () => {
    const eylem = homeAction(null);
    expect(eylem.clearUrl).toBe(false);
    expect(visibleCoords(eylem.sel, null, EV)).toEqual(EV);
  });

  /**
   * ⭐ `sel` daima `null` — eve ait koordinat SABİTLENMEZ. Oyuncu sonradan şehir değiştirirse
   * görünüm yeni aktif şehri izlemeli; koordinatı `sel`e yazsaydık takip ölürdü.
   */
  it('seçimi sabitlemez, aktif şehri izlemeye döner', () => {
    expect(homeAction(RAPOR).sel).toBeNull();
    const yeniEv: Coords = { k: 9, d: 400 };
    expect(visibleCoords(homeAction(RAPOR).sel, null, yeniEv)).toEqual(yeniEv);
  });
});

File: apps/mobile/test/core/world_coords_test.dart
/// DÜNYA KOORDİNAT ÖNCELİĞİ — web'de **canlı bir hataya** yol açan kural.
///
/// ⚠️ Hata (kullanıcı, 2026-08-16): «Kendi diyarıma dön» düğmesi yalnız seçimi boşaltıyordu ve
/// derin bağlantıyla açılmış sayfada sıra **eve değil adrese** düşüyordu — casusluk raporundan
/// gelen oyuncu kendi diyarına değil raporun diyarına dönüyordu. Aşağıdaki son grup tam olarak
/// bunu kilitliyor.
library;

import 'package:flutter_test/flutter_test.dart';
import 'package:mobilwar/core/world_coords.dart';

void main() {
  const ev = (k: 2, d: 40);
  const adres = (k: 5, d: 120);
  const secim = (k: 7, d: 9);

  group('visibleCoords — öncelik zinciri', () {
    test('hiçbiri yoksa 1:1', () {
      expect(visibleCoords(null, null, null), (k: 1, d: 1));
    });

    test('yalnız ev varsa eve bakılır', () {
      expect(visibleCoords(null, null, ev), ev);
    });

    /// ⚠️ Adres evin ÖNÜNDE olmak zorunda: derin bağlantıyla gelen oyuncu ilk açılışta
    /// hedefi görmeli, kendi diyarını değil.
    test('adres evi EZER', () {
      expect(visibleCoords(null, adres, ev), adres);
    });

    test('elle seçim her şeyi ezer', () {
      expect(visibleCoords(secim, adres, ev), secim);
    });

    /// ⚠️ `k` ve `d` AYRI çözülüyor: oyuncu yalnız kıtayı değiştirdiğinde diyar numarası
    /// yerinde kalmalı. Çift olarak çözülseydi kıta seçimi diyarı da sıfırlardı.
    test('k ve d ayrı ayrı düşüyor', () {
      expect(visibleCoords((k: 7, d: 9), null, ev), (k: 7, d: 9));
      // Yalnız d taşıyan bir seçim: k eve düşmeli.
      const yalnizD = (k: 3, d: 9);
      expect(visibleCoords(yalnizD, null, ev).k, 3);
    });
  });

  group('homeAction — «kendi diyarıma dön»', () {
    /// ⚠️⚠️ Hatanın ta kendisi: adres varken seçimi boşaltmak YETMEZ, adres de bırakılmalı.
    test('adres varsa adres de bırakılıyor', () {
      final a = homeAction(adres);
      expect(a.sel, isNull);
      expect(a.clearUrl, isTrue);
    });

    test('adres yoksa yalnız seçim bırakılıyor', () {
      final a = homeAction(null);
      expect(a.sel, isNull);
      expect(a.clearUrl, isFalse);
    });

    /// Düğmenin sonucunu UÇTAN UCA ölçüyor: seçim ve adres bırakılınca görünüm EVE düşmeli.
    test('⭐ düğmeden sonra görünüm gerçekten eve dönüyor', () {
      final a = homeAction(adres);
      final sonra = visibleCoords(a.sel, a.clearUrl ? null : adres, ev);
      expect(sonra, ev);
    });

    /// ⚠️ Eve ait koordinat `sel`e YAZILMIYOR: `sel = null` "aktif şehri izle" demek.
    /// Oyuncu sonra şehir değiştirirse görünüm onu takip etmeli.
    test('⭐ eve döndükten sonra şehir değişimi takip ediliyor', () {
      final a = homeAction(null);
      const yeniEv = (k: 8, d: 300);
      expect(visibleCoords(a.sel, null, yeniEv), yeniEv);
    });
  });

  group('realmFromPath', () {
    test('geçerli adres okunuyor', () {
      expect(realmFromPath('5', '120'), (k: 5, d: 120));
    });

    /// ⚠️ Kelepçeleme ATMA DEĞİL: bozuk bağlantı oyuncuyu sessizce eve atmamalı.
    test('sınır dışı adres kırpılıyor, reddedilmiyor', () {
      expect(realmFromPath('11', '900'), (k: 10, d: 500));
    });

    test('sayı olmayan ya da sıfır/negatif adres null', () {
      expect(realmFromPath('a', '1'), isNull);
      expect(realmFromPath('1', ''), isNull);
      expect(realmFromPath('0', '5'), isNull);
      expect(realmFromPath('-3', '5'), isNull);
      expect(realmFromPath(null, null), isNull);
    });
  });
}
